lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -0,0 +1,837 @@
+            }
+
+            // re-create without overwrite is an error
+            assertThatThrownBy(outputFile::create)
+                    .isInstanceOf(FileAlreadyExistsException.class)
+                    .hasMessageContaining(tempBlob.location());
+
+            // verify nothing changed
+            assertThat(tempBlob.read()).isEqualTo("initial");
+
+            // overwrite file
+            try (OutputStream outputStream = outputFile.createOrOverwrite()) {
+                outputStream.write("overwrite".getBytes(UTF_8));
+            }
+
+            // verify file is different
+            assertThat(tempBlob.read()).isEqualTo("overwrite");
+        }
+    }
+
+    @Test
+    void testOutputStreamByteAtATime()
+            throws IOException
+    {
+        try (TempBlob tempBlob = randomBlobLocation("inputStream")) {
+            try (OutputStream outputStream = tempBlob.outputFile().create()) {
+                for (int i = 0; i < MEGABYTE; i++) {
+                    outputStream.write(i);
+                    if (i % 1024 == 0) {
+                        outputStream.flush();
+                    }
+                }
+                outputStream.close();
+
+                // verify all the methods throw after close
+                assertThatThrownBy(() -> outputStream.write(42))
+                        .isInstanceOf(IOException.class)
+                        .hasMessageContaining(tempBlob.location());
+                assertThatThrownBy(() -> outputStream.write(new byte[10]))
+                        .isInstanceOf(IOException.class)
+                        .hasMessageContaining(tempBlob.location());
+                assertThatThrownBy(() -> outputStream.write(new byte[10], 1, 3))
+                        .isInstanceOf(IOException.class)
+                        .hasMessageContaining(tempBlob.location());
+                assertThatThrownBy(outputStream::flush)
+                        .isInstanceOf(IOException.class)
+                        .hasMessageContaining(tempBlob.location());
+            }
+
+            try (TrinoInputStream inputStream = tempBlob.inputFile().newStream()) {
+                for (int i = 0; i < MEGABYTE; i++) {
+                    int value = inputStream.read();
+                    assertThat(value).isGreaterThanOrEqualTo(0);
+                    assertThat((byte) value).isEqualTo((byte) i);
+                }
+            }
+        }
+    }
+
+    @Test
+    void testPaths()
+            throws IOException
+    {
+        if (isHierarchical()) {
+            testPathHierarchical();
+        }
+        else {
+            testPathBlob();
+        }
+    }
+
+    protected void testPathHierarchical()
+            throws IOException
+    {
+        // file outside of root is not allowed
+        assertThatThrownBy(() -> getFileSystem().newOutputFile(createLocation("../file")).createOrOverwrite().close())
+                .isInstanceOf(Exception.class)
```

Comment:
`Exception` seems too general here. Where is this check for the memory implementation?